walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,103 @@
+      * for just optional attendees.
+      */
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        long meetingDuration = request.getDuration();
+        // Required attendees for the request meeting.
+        Collection<String> requiredAttendees = request.getAttendees(); 
+        // All attendees (required and optional) for the request meeting.
+        Collection<String> allAttendees = new ArrayList<>(requiredAttendees);
+        allAttendees.addAll(request.getOptionalAttendees());
+
+        // If the meeting duration is invalid, return an empty Collection.
+        if (meetingDuration > WHOLE_DAY.duration() || meetingDuration == 0) {
+            return Arrays.asList();
+        }
+
+        // Find available TimeRange for both required and optional attendees.
+        Collection<TimeRange> allMeetingTimes = 
+            findMeetingTimes(events, allAttendees, meetingDuration);
+
+        // If there is meeting time for all, return allMeetingTimes so that 
+        // both required and optional can attend the meeting. If there is no 
+        // required attendee, return allMeetingTimes for optional attendees.
+        if (allMeetingTimes.size() > 0 || requiredAttendees.size() == 0) {
+            return allMeetingTimes;
+        }
+        // If not, return available TimeRange for required attendees.
+        return findMeetingTimes(
+            events, requiredAttendees, meetingDuration);
+    }
+
+    /**
+      * Find all times that the meeting could happen, given attendees and duration. 
+      * Pointer rangeStart is pointing at the start of the current available 
+      * TimeRange to keep track of where the meeting could possibly start. The time 
+      * between the rangeStart and start time of the current event is the current 
+      * avaliable TimeRange. If the current available TimeRange has longer duration 
+      * than the requested meeting duration, add it to available meetingTimes.
+      */
+    private static Collection<TimeRange> findMeetingTimes(
+        Collection<Event> events, Collection<String> attendees, long duration) {
+        Collection<TimeRange> meetingTimes = new ArrayList<TimeRange>();
+
+        // Sort the events by the start time.
+        List<Event> eventsSorted = new ArrayList<Event>(events);
+        Collections.sort(eventsSorted, Event.SORT_BY_START_TIME);
+
+        int rangeStart = DAY_START; // Pointer to start of the previous event.
+
+        // Linear search the current events and add new TimeRange to 
+        // meetingTimes.
+        for (Event event : eventsSorted) {
+            Collection<String> eventAttendees = event.getAttendees();
+            TimeRange eventWhen = event.getWhen();
+            int eventStart = eventWhen.start();
+            int eventEnd = eventWhen.end();
+
+            // Check if given attendees shows up in the event.
+            if (attendees.containsAll(eventAttendees)) {
```

Comment:
This doesn't seem correct. It is true only when eventAttendees is a subset of input attendees. Shouldn't it be true as long as *one of* eventAttendees is in input attendees?

Reply:
> The containsAll() method of Java Set is used to check whether two sets contain the same elements or not. It takes one set as a parameter and returns True if all of the elements of this set is present in the other set. 

I think it is not checking for the subset.

Reply:
> It takes one set as a parameter and returns True if all of the elements of this set is present in the other set

That's exactly what I meant by subset.

Sorry I still didn't get it. For example:
```
attendees = {A, B}
events = {
   // event_name : attendees
   x: A, D;
   y: B, D;
   z: A, B;
    }
```
Only event 'z' will fall into the if clause while events 'x' and 'y' won't even be checked because their attendees (A, D and B, D) are not a subset of attendees(A, B).
Am I missing anything?

Reply:
Oh I see! It is a bug in my code. I think the testing and example I had all had one attendee for each event, so I just let it slip. Thanks for pointing it out! :)